### ISSUE TITLE
DB::Exception: inflate failed: buffer error

### ISSUE THREAD

**Issue by mindprince** [yandex/ClickHouse#412]
Hi,  when inserting to Clickhouse with HTTP transport and gzip (compression level 1), I see a lot of stack traces like the following. They don't happen regularly (sometimes the pipeline keeps running without errors for a long time), I see these errors usually when the size of the gzipped data reaches around 100MB+.

There are three different stack traces which are common:
```
2017.01.27 00:59:40.444106 [ 10 ] <Error> executeQuery: Code: 354, e.displayText() = DB::Exception: inflate failed: buffer error: (at row 47486), e.what() = DB::Exception (from [::ffff:10.36.14.3]:52380) (in query: INSERT INTO r0.some_table_replica FORMAT TabSeparated  ), Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x122ad76]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::string const&, int)+0x1f) [0xf8792f]
2. /usr/bin/clickhouse-server(DB::ZlibInflatingReadBuffer::nextImpl()+0x277) [0x1240537]
3. /usr/bin/clickhouse-server(DB::ConcatReadBuffer::nextImpl()+0x3e) [0xf8fc5e]
4. /usr/bin/clickhouse-server(DB::ConcatReadBuffer::nextImpl()+0x3e) [0xf8fc5e]
5. /usr/bin/clickhouse-server(DB::TabSeparatedRowInputStream::read(DB::Block&)+0x31c) [0x25f80cc]
6. /usr/bin/clickhouse-server(DB::BlockInputStreamFromRowInputStream::readImpl()+0x3d) [0x261449d]
7. /usr/bin/clickhouse-server(DB::IProfilingBlockInputStream::read()+0x2a0) [0x10d3060]
8. /usr/bin/clickhouse-server(DB::InputStreamFromASTInsertQuery::readImpl()+0x18) [0x119dfb8]
9. /usr/bin/clickhouse-server(DB::IProfilingBlockInputStream::read()+0x2a0) [0x10d3060]
10. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0x5c) [0x1d300fc]
11. /usr/bin/clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, DB::Context&, std::shared_ptr<DB::IBlockInputStream>&, std::function<void (std::string const&)>)+0x285) [0x119d8e5]
12. /usr/bin/clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x10de) [0xf8e10e]
13. /usr/bin/clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x222) [0xf8f812]
14. /usr/bin/clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x27b) [0x305e74b]
15. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3042e5f]
16. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x10b) [0x30609eb]
17. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x87) [0x30d8707]
18. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x96) [0x309c086]
19. /lib/x86_64-linux-gnu/libpthread.so.0(+0x80a4) [0x7f7b51eaa0a4]
20. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f7b514d262d]
```

```
2017.01.27 19:15:59.631346 [ 70 ] <Error> HTTPHandler: Code: 354, e.displayText() = DB::Exception: inflate failed: buffer error, e.what() = DB::Exception, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x122ad76]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::string const&, int)+0x1f) [0xf8792f]
2. /usr/bin/clickhouse-server(DB::ZlibInflatingReadBuffer::nextImpl()+0x277) [0x1240537]
3. /usr/bin/clickhouse-server(DB::ConcatReadBuffer::nextImpl()+0x3e) [0xf8fc5e]
4. /usr/bin/clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, DB::Context&, std::shared_ptr<DB::IBlockInputStream>&, std::function<void (std::string const&)>)+0x6fc) [0x119dd5c]
5. /usr/bin/clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x10de) [0xf8e10e]
6. /usr/bin/clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x222) [0xf8f812]
7. /usr/bin/clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x27b) [0x305e74b]
8. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3042e5f]
9. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x10b) [0x30609eb]
10. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x87) [0x30d8707]
11. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x96) [0x309c086]
12. /lib/x86_64-linux-gnu/libpthread.so.0(+0x80a4) [0x7fac06e5d0a4]
13. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7fac0648562d]
```

```
2017.01.27 13:32:10.967702 [ 29 ] <Error> HTTPHandler: Code: 354, e.displayText() = DB::Exception: inflate failed: data error, e.what() = DB::Exception, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x122ad76]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::string const&, int)+0x1f) [0xf8792f]
2. /usr/bin/clickhouse-server(DB::ZlibInflatingReadBuffer::nextImpl()+0x277) [0x1240537]
3. /usr/bin/clickhouse-server(DB::ConcatReadBuffer::nextImpl()+0x3e) [0xf8fc5e]
4. /usr/bin/clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, DB::Context&, std::shared_ptr<DB::IBlockInputStream>&, std::function<void (std::string const&)>)+0x6fc) [0x119dd5c]
5. /usr/bin/clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x10de) [0xf8e10e]
6. /usr/bin/clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x222) [0xf8f812]
7. /usr/bin/clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x27b) [0x305e74b]
8. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3042e5f]
9. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x10b) [0x30609eb]
10. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x87) [0x30d8707]
11. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x96) [0x309c086]
12. /lib/x86_64-linux-gnu/libpthread.so.0(+0x80a4) [0x7fac06e5d0a4]
13. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7fac0648562d]
```

**Comment by alexey-milovidov** [comment]
Hi. Thank you for the report!
Does it reproduce if you change compression level to 2?

There could be several reasons:

1. We recently have changed usual zlib library to [zlib-ng](https://github.com/Dead2/zlib-ng). It could be related. Try to get reproducible test case: could you create one file, which always triggers the error?
Then we will try to debug. Even if file will contain confidential data, we could debug by trying different variants of library on your side, without disclosure of data.

2. Broken network connection. In this case, you should look at error message, that you get on client side.

**Comment by alexey-milovidov** [comment]
And also quick question. If we will add support of Brotli (`Content-Encoding: br`), it will be more efficient than gzip both in terms of compression ratio and compression speed. Do you appreciate?

**Comment by ztlpn** [comment]
Hi, a couple more questions: how do you create your compressed data? Is it with stock zlib? Can you make sure that the stream is indeed correct (i.e. stock gzip decompresses it without problems)? There are bugs in level 1 compression algorithm in some of the zlib forks floating around (including zlib-ng and Intel zlib) so it makes sense to first ensure that the stream is correct.

**Comment by vavrusa** [comment]
We use [cgzip](https://github.com/youtube/vitess/tree/master/go/cgzip), not sure what gzip lib is it linked against but it's possible it's broken only with the best speed compression level. It's working fine with [v1.1.54127-stable](https://github.com/yandex/ClickHouse/releases/tag/v1.1.54127-stable)

**Comment by ztlpn** [comment]
> It's working fine with v1.1.54127-stable

That means the compressed data is OK and the problem is either with zlib-ng or our code. I'm going to create a branch with zlib-ng substituted for stock zlib after the weekend. It would be great if you'll be able to build it and test against it.

**Comment by mindprince** [comment]
> Does it reproduce if you change compression level to 2?

I deployed inserters with compression level 2, got the same error (v1.1.54134-stable)

```
2017.01.29 22:44:36.505169 [ 352 ] <Error> executeQuery: Code: 354, e.displayText() = DB::Exception: inflate failed: buffer error: (at row 221474), e.what() = DB::Exception (from [::ffff:10.36.11.12]:51294) (in query: INSERT INTO r0.requests_error_replica FORMAT TabSeparated  ), Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x122ad76]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::string const&, int)+0x1f) [0xf8792f]
2. /usr/bin/clickhouse-server(DB::ZlibInflatingReadBuffer::nextImpl()+0x277) [0x1240537]
3. /usr/bin/clickhouse-server(DB::ConcatReadBuffer::nextImpl()+0x3e) [0xf8fc5e]
4. /usr/bin/clickhouse-server(DB::ConcatReadBuffer::nextImpl()+0x3e) [0xf8fc5e]
5. /usr/bin/clickhouse-server(DB::TabSeparatedRowInputStream::read(DB::Block&)+0x31c) [0x25f80cc]
6. /usr/bin/clickhouse-server(DB::BlockInputStreamFromRowInputStream::readImpl()+0x3d) [0x261449d]
7. /usr/bin/clickhouse-server(DB::IProfilingBlockInputStream::read()+0x2a0) [0x10d3060]
8. /usr/bin/clickhouse-server(DB::InputStreamFromASTInsertQuery::readImpl()+0x18) [0x119dfb8]
9. /usr/bin/clickhouse-server(DB::IProfilingBlockInputStream::read()+0x2a0) [0x10d3060]
10. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0x5c) [0x1d300fc]
11. /usr/bin/clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, DB::Context&, std::shared_ptr<DB::IBlockInputStream>&, std::function<void (std::string const&)>)+0x285) [0x119d8e5]
12. /usr/bin/clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x10de) [0xf8e10e]
13. /usr/bin/clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x222) [0xf8f812]
14. /usr/bin/clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x27b) [0x305e74b]
15. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3042e5f]
16. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x10b) [0x30609eb]
17. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x87) [0x30d8707]
18. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x96) [0x309c086]
19. /lib/x86_64-linux-gnu/libpthread.so.0(+0x80a4) [0x7f05271720a4]
20. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f052679a62d]
```

**Comment by ztlpn** [comment]
Could you try to reproduce with the version compiled from [this branch](https://github.com/yandex/ClickHouse/tree/original-zlib)?

Meanwhile I'll try to reproduce the issue on our servers.

**Comment by mindprince** [comment]
I will try this out very soon.

**Comment by proller** [comment]
Also in current master now possible to build with -DUSE_INTERNAL_ZLIB_LIBRARY=0
it will use your system zlib

**Comment by mindprince** [comment]
So, I had been putting off reproducing this for sometime (as we started running our inserters without compression and it wasn't hurting.)

However, during the weekend, while debugging some other issue I realized that this bug was NOT a problem with ClickHouse or zlib. It was a bug in our inserter. When our inserter is unable to insert to one Clickhouse node it tries to insert to another node and so on. This works fine when the insertion failure happens before we send any data. Sometimes, however, the insertion failure happens after we have sent some data because of broken streams etc. In that case, we were not resetting our buffers and sending a bad stream which was causing the  "inflate failed: buffer error" at the server.

The broken streams were only happening when the stream size was big, that's why we were not seeing it regularly and only seeing it when the gzipped stream size was 100MB+. That is also the reason why we were not seeing it with the v1.1.54127-stable version, we have two ClickHouse clusters running, the one running v1.1.54127-stable never reached 100MB+ gzipped streams while inserting.

Thanks a lot for looking into this issue guys. Really sorry for wasting your time on this. :(
